cranelift/codegen/src/egraph/elaborate.rs

```diff
@@ -278,6 +278,14 @@ impl<'a> Elaborator<'a> {
                         before
                     );
 
+                    // Get the best option; we use `value` (latest
+                    // value) here so we have a full view of the
+                    // eclass.
+                    trace!("looking up best value for {}", value);
+                    let (_, best_value) = self.value_to_best_value[value];
+                    debug_assert_ne!(best_value, Value::reserved_value());
+                    trace!("elaborate: value {} -> best {}", value, best_value);
+
                     let remat = if let Some(elab_val) =
                         self.value_to_elaborated_value.get(&canonical_value)
```

Comment:
Should this lookup also switch to `best_value` instead of `canonical_value`? What about the later uses of `canonical_value`? Perhaps `canonical_value` should just be dead as soon as we've looked up `best_value`, maybe enforced by shadowing it?

Reply:
This actually needs to be the canonical-value, to avoid unnecessary codegen!

The reason is that we record the elaborated values in the map with canonical-ID as key, and we want to dedup as much as possible here. Otherwise if we refer to a value with different IDs that point to different parts of the eclass, we'll have unnecessary misses and recomputations.

However it's good that you asked because I went over the whole function and saw that below, another `canonical_value` should be `best_value`, when we look up remat in the other case!